File: ts/src/Zone.ts
import Counter from "./Counter";

export default class Zone {
  name: string;
  counters: Counter[];
  occupancy: number;
  constructor(name: string, counters: Counter[]) {
    this.name = name;
    this.counters = counters;
    this.occupancy = 0;
  }
  update() {
    let totalIn = 0;
    let totalOut = 0;
    for (let counter of this.counters) {
      counter.update();
      totalIn += counter.totalIn;
      totalOut += counter.totalOut;
    }

    this.occupancy = totalIn - totalOut;
  }
}

File: elixir/lib/zone.ex
defmodule Zone do
  def new(name, counters) do
    %{name: name, counters: counters, occupancy: 0}
  end

  def update(state) do
    counters = Enum.map(state.counters, fn counter -> Counter.update(counter) end)
    total_in = Enum.reduce(state.counters, 0, fn counter, acc ->
      acc + counter.total_in
    end)
    total_out = Enum.reduce(state.counters, 0, fn counter, acc ->
      acc + counter.total_out
    end)
    %{state | occupancy: total_in - total_out, counters: counters}
  end

  def occupancy(state), do: state.occupancy
end
